llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -1207,6 +1207,34 @@ Instruction *InstCombinerImpl::commonIDivTransforms(BinaryOperator &I) {
     }
   }
 
+  // (X * Y) / (X * Z) --> Y / X (and commuted variants)
+  if (match(Op0, m_Mul(m_Value(X), m_Value(Y)))) {
+    auto OB0HasNSW = cast<OverflowingBinaryOperator>(Op0)->hasNoSignedWrap();
+    auto OB0HasNUW = cast<OverflowingBinaryOperator>(Op0)->hasNoUnsignedWrap();
+
+    auto CreateDivOrNull = [&](Value *A, Value *B) -> Instruction * {
+      auto OB1HasNSW = cast<OverflowingBinaryOperator>(Op1)->hasNoSignedWrap();
+      auto OB1HasNUW =
+          cast<OverflowingBinaryOperator>(Op1)->hasNoUnsignedWrap();
+      const APInt *C;
+      if (IsSigned && OB0HasNSW && OB1HasNSW && (match(B, m_APInt(C))) &&
+          (*C != -1))
+        return BinaryOperator::CreateSDiv(A, B);
+      if (!IsSigned && OB0HasNUW && OB1HasNUW) {
+        return BinaryOperator::CreateUDiv(A, B);
+      };
+      return nullptr;
```

Comment:
Possible implementation:
```suggestion
      if (!IsSigned && OB0HasNUW) {
         if (auto *CV = dyn_cast<Constant>(simplifyUDivInst(A, B, /*IsExact*/ true, SQ.getWithInstruction(&I)));
         CV && !(isa<PoisonValue>(CV) || CV->containsPoisonElement()))
             return CV;
      }
      return nullptr;
```